crates/turbopack/src/lib.rs

```diff
@@ -259,113 +259,129 @@ impl ModuleAssetContextVc {
-
-        let reference_type = reference_type.into_value();
-        let part = match &reference_type {
-            ReferenceType::EcmaScriptModules(EcmaScriptModulesReferenceSubType::ImportPart(
-                part,
-            )) => Some(*part),
-            _ => None,
-        };
-        let mut current_source = source;
-        let mut current_module_type = None;
-        for rule in options.await?.rules.iter() {
-            if rule
-                .matches(source, &*ident.path().await?, &reference_type)
-                .await?
-            {
-                for effect in rule.effects() {
-                    match effect {
-                        ModuleRuleEffect::SourceTransforms(transforms) => {
-                            current_source = transforms.transform(current_source);
-                            if current_source.ident().resolve().await? != ident {
-                                // The ident has been changed, so we need to apply new rules.
-                                return Ok(self_vc
-                                    .process_default(current_source, Value::new(reference_type)));
-                            }
-                        }
-                        ModuleRuleEffect::ModuleType(module) => {
-                            current_module_type = Some(*module);
+    ) -> AssetVc {
+        process_default(self_vc, source, reference_type, Vec::new())
+    }
+}
+
+#[turbo_tasks::function]
+async fn process_default(
+    context: ModuleAssetContextVc,
+    source: AssetVc,
+    reference_type: Value<ReferenceType>,
+    processed_rules: Vec<usize>,
+) -> Result<AssetVc> {
+    let ident = source.ident().resolve().await?;
+    let options = ModuleOptionsVc::new(ident.path().parent(), context.module_options_context());
+
+    let reference_type = reference_type.into_value();
+    let part = match &reference_type {
+        ReferenceType::EcmaScriptModules(EcmaScriptModulesReferenceSubType::ImportPart(part)) => {
+            Some(*part)
+        }
+        _ => None,
+    };
+    let mut current_source = source;
+    let mut current_module_type = None;
+    for (i, rule) in options.await?.rules.iter().enumerate() {
+        if processed_rules.contains(&i) {
+            continue;
+        }
+        if rule
+            .matches(source, &*ident.path().await?, &reference_type)
+            .await?
+        {
+            for effect in rule.effects() {
+                match effect {
+                    ModuleRuleEffect::SourceTransforms(transforms) => {
+                        current_source = transforms.transform(current_source);
+                        if current_source.ident().resolve().await? != ident {
+                            // The ident has been changed, so we need to apply new rules.
+                            let mut processed_rules = processed_rules.clone();
+                            processed_rules.push(i);
```

Comment:
Can you leave a comment explaining this? I think this is to avoid an infinite loop for `foo.js` -> `foo.transformed.js` -> `foo.transformed.transformed.js`?

If we only want to exclude the transformer that _just_ ran, we could just use an index instead of a `Vec`.

Reply:
You could build a cycle with multiple items:

```
"*.a": { as: "*.b" },
"*.b": { as: "*.c" },
"*.c": { as: "*.a" },
```

But it also handles valid cases like:

```
"*.optimize.*": { loader: "image-optimizer", as: "*" },
"*.minimize.*": { loader: "minimize-loader", as: "*" },
"*.svg": { loader: "svg-loader" }
```

`file.optimize.minimize.svg` need to have multiple `processed_rules`